exec/java-exec/src/main/java/org/apache/drill/exec/rpc/user/UserSession.java

```diff
@@ -207,18 +245,85 @@ public SchemaPlus getDefaultSchema(SchemaPlus rootSchema) {
       return null;
     }
 
-    final SchemaPlus defaultSchema = SchemaUtilites.findSchema(rootSchema, defaultSchemaPath);
+    return SchemaUtilites.findSchema(rootSchema, defaultSchemaPath);
+  }
 
-    if (defaultSchema == null) {
-      // If the current schema resolves to null, return root schema as the current default schema.
-      return defaultSchema;
+  public boolean setSessionOption(String name, String value) {
+    return true;
+  }
+
+  /**
+   * @return unique session identifier
+   */
+  public String getSessionId() { return sessionId; }
+
+  /**
+   * Creates and adds session temporary location if absent using schema configuration.
+   * Generates temporary table name and stores it's original name as key
+   * and generated name as value in  session temporary tables cache.
+   * If original table name already exists, new name is not regenerated and is reused.
+   * This can happen if default temporary workspace was changed (file system or location) or
+   * orphan temporary table name has remained (name was registered but table creation did not succeed).
+   *
+   * @param schema table schema
+   * @param tableName original table name
+   * @return generated temporary table name
+   * @throws IOException if error during session temporary location creation
+   */
+  public String registerTemporaryTable(AbstractSchema schema, String tableName) throws IOException {
+    if (schema instanceof WorkspaceSchemaFactory.WorkspaceSchema) {
+      addTemporaryLocation((WorkspaceSchemaFactory.WorkspaceSchema) schema);
+      String temporaryTableName = Paths.get(sessionId, UUID.randomUUID().toString()).toString();
+      String oldTemporaryTableName = temporaryTables.putIfAbsent(tableName, temporaryTableName);
+      return oldTemporaryTableName == null ? temporaryTableName : oldTemporaryTableName;
+    } else {
+      throw new DrillRuntimeException("Temporary workspace must be instance of WorkspaceSchemaFactory.WorkspaceSchema");
```

Comment:
Should this check be done earlier? Say, on start up? Otherwise, the admin won't know about the problem until some random time after each Drillbit starts. Also, since this is a config problem, shouldn't we log this at the error level so that the admin can track down the problem? Telling the query user is fine, but the user is not in a position to fix the problem.